### PR TITLE
[RHELC-1003] Allow to skip kernel check even if packages are available

### PR DESCRIPTION
**Pull request by wombelix**
The environment variable `CONVERT2RHEL_UNSUPPORTED_SKIP_KERNEL_CURRENCY_CHECK` doesn't work as expected in my opinion. The kernel package check will only be skipped if no kernel packages are found in the repo. If packages are available, the env var will be ignored and the conversion fails.

In a rare edge case, were a new minor kernel release in RHEL 7 is already available but not in CentOS 7, forcing the check to be skipped is the easiest way to move forward without jumping through further hoops. Observed with `kernel-3.10.0-1160.90.1.el7` as latest in RHEL 7 and `kernel-3.10.0-1160.88.1.el7` as latest in CentOS 7.

Even with set env var, the kernel check will not be skipped in `convert2rhel-1.2-2.el7.noarch`:
```
[05/02/2023 19:17:18] TASK - [Prepare: Check if the loaded kernel version is the most recent] ***
CRITICAL - The version of the loaded kernel is different from the latest version in the enabled system repositories.
 Latest kernel version available in rhel-7-server-rhui-rpms: 3.10.0-1160.90.1.el7
 Loaded kernel version: 3.10.0-1160.88.1.el7

To proceed with the conversion, update the kernel version by executing the following step:

1. yum install kernel-3.10.0-1160.90.1.el7 -y
2. reboot
Writing breadcrumbs to '/etc/migration-results'.
Writing RHSM custom facts to '/etc/rhsm/facts/convert2rhel.facts'.
No changes were made to the system.

[root@ip-10-0-138-67 etc]# printenv |grep -i convert2rhel
CONVERT2RHEL_UNSUPPORTED_SKIP_KERNEL_CURRENCY_CHECK=true
```

As follow up issue from skipping the kernel check, the kernel module check also fails:
```
CRITICAL - The following loaded kernel modules are not available in RHEL:
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/ghash-clmulni-intel.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/net/ipv4/netfilter/ip_tables.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/acpi/nfit/nfit.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/crct10dif-pclmul.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/crypto/lrw.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/crc32-pclmul.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/net/sunrpc/sunrpc.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/glue_helper.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/lib/libcrc32c.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/parport/parport_pc.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/input/misc/pcspkr.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/crypto/cryptd.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/input/serio/serio_raw.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/i2c/busses/i2c-piix4.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/platform/intel/iosf_mbi.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/parport/parport.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/crypto/gf128mul.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/aesni-intel.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/nvme/host/nvme-core.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/nvdimm/libnvdimm.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/fs/xfs/xfs.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/net/ethernet/amazon/ena/ena.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/char/ppdev.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/ablk_helper.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/drivers/nvme/host/nvme.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/crypto/crct10dif_common.ko.xz
/lib/modules/3.10.0-1160.88.1.el7.x86_64/kernel/arch/x86/crypto/crc32c-intel.ko.xz
Ensure you have updated the kernel to the latest available version and rebooted the system.
If this message persists, you can prevent the modules from loading by following https://access.redhat.com/solutions/41278 and rerun convert2rhel.
Keeping them loaded could cause the system to malfunction after the conversion as they might not work properly with the RHEL kernel.
To circumvent this check and accept the risk you can set environment variable 'CONVERT2RHEL_ALLOW_UNAVAILABLE_KMODS=1'.
Writing breadcrumbs to '/etc/migration-results'.
Writing RHSM custom facts to '/etc/rhsm/facts/convert2rhel.facts'.
WARNING - Abnormal exit! Performing rollback ...
```

But this check could then be skipped with `CONVERT2RHEL_ALLOW_UNAVAILABLE_KMODS=1` as expected.

I'm aware that skipping checks can cause massive issues, but there might be cases were someone really want it and needs an easy way to do that.

Jira Issues: [RHELC-1003](https://issues.redhat.com/browse/RHELC-1003)

Checklist

- [x] PR has been tested manually in a VM (either author or reviewer)
- [x] Jira issue has been made public if possible
- [x] `[RHELC-]` is part of the PR title 
- [ ] GitHub label has been added to help with Release notes 
- [x] PR title explains the change from the user's point of view
- [x] Code and tests are documented properly
- [x] The commits are squashed to as few commits as possible (without losing data) 
- [ ] When merged: Jira issue has been updated to `Release Pending` if relevant

### CODE REVIEW

**Comment by bocekm** [comment]
Thanks, @wombelix!

Agreed, the skip env var should work as you outlined. We have bumped into this issue as well and we're tracking it under https://issues.redhat.com/browse/RHELC-884.

The output you posted shows that you had the `rhel-7-server-rhui-rpms` repository enabled on the CentOS Linux 7 system. The convert2rhel tool does not expect the RHEL repos to be enabled prior to the conversion as it is the job of the tool to enable the right RHEL repos during the conversion. But the fact is the conversion should not blocked by the kernel check.

The tool does not support using RHUI for accessing RHEL content so I suppose you're trying to make the conversion work with it somehow.

**Comment by wombelix** [comment]
> Agreed, the skip env var should work as you outlined. We have bumped into this issue as well and we're tracking it under https://issues.redhat.com/browse/RHELC-1003.
> 

Cool, glad it will be fixed, surprised to see the checks failing, I did local tests as described in the contribution doc and a convert2rhel test run were the step was skipped as expected.

I have to check your commit later to see how you improved / fixed it.

> The output you posted shows that you had the `rhel-7-server-rhui-rpms` repository enabled on the CentOS Linux 7 system. The convert2rhel tool does not expect the RHEL repos to be enabled prior to the conversion as it is the job of the tool to enable the right RHEL repos during the conversion. But definitely the conversion should not blocked by the kernel check this way.
> 

They should be activated because I run `convert2rhel --debug --no-rhsm --enablerepo=rhel-7-server-rhui-rpms --enablerepo=rhel-7-server-rhui-rh-common-rpms --enablerepo=rhui-client-config-server-7`, before they were actually disabled.

> The tool does not support using RHUI for accessing RHEL content so I suppose you're trying to make the conversion work with it somehow.

I'm aware that it's not officially supported (yet), it's part of a POC @nlwmn is working on to help AWS Customers to migrate CentOS 7 Systems to RHEL 7 PAYG. Manually and with Ansible playbooks Nicolas wrote, we did a lot successful conversions during the last month already.

So RHUI doesn't seem to be our problem, at least it wasn't as long there was no diff in the released package versions in the RHEL and CentOS repos, which feels like a rare edge case, most of the time everything should run smoothly.

**Comment by danmyway** [comment]
Rebased to main and added a test case to verify the code changes.

**Review by kokesak — COMMENTED** [review]
Few things, otherwise looks good :D

File: tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        )
+    ) as c2r:
+        # Verify that using the deprecated environment variable is still allowed and continues the conversion
+        # TODO(danmyway) uncomment in #684
+        # assert c2r.expect("You are using the deprecated 'CONVERT2RHEL_UNSUPPORTED_SKIP_KERNEL_CURRENCY_CHECK'") == 0
+        # Make sure the kernel comparison is skipped
+        c2r_expect_index = c2r.expect(
+            [
+                "we will skip the kernel comparison.",
+                "Could not find any kernel packages in repositories to compare against the loaded kernel.",
```
Comment (kokesak) [inline tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:46]: This will probably timeots if it doesn't find either of those values right?

**Comment by bocekm** [comment]
> > The output you posted shows that you had the `rhel-7-server-rhui-rpms` repository enabled on the CentOS Linux 7 system. The convert2rhel tool does not expect the RHEL repos to be enabled prior to the conversion as it is the job of the tool to enable the right RHEL repos during the conversion. But definitely the conversion should not blocked by the kernel check this way.
> 
> They should be activated because I run `convert2rhel --debug --no-rhsm --enablerepo=rhel-7-server-rhui-rpms --enablerepo=rhel-7-server-rhui-rh-common-rpms --enablerepo=rhui-client-config-server-7`, before they were actually disabled.

@abadger, you've been investing at [an issue](https://github.com/oamg/convert2rhel/pull/818) that having or not having certain repos in yum cache may result in weird repoquery behavior.

I have a hunch that the `yum makecache` we run only for custom repos in [`check_custom_repos_are_valid()`](https://github.com/oamg/convert2rhel/blob/v1.2/convert2rhel/checks.py#L334) may cause this issue @wombelix is facing. The only reason of running the makecache is verifying that the repos are accessible.

When [this repoquery command](https://github.com/oamg/convert2rhel/blob/v1.2/convert2rhel/checks.py#L722) is being executed, it should be looking **only** at repos **enabled** on the system. @wombelix had them disabled though, prior to the conversion. It seems to me that because we created the yum cache for those disabled custom repos, the repoquery is now shamelessly using the cache.

If that's so, then a quick solution would be to clean the cache right after the repo validation.

> 
> > The tool does not support using RHUI for accessing RHEL content so I suppose you're trying to make the conversion work with it somehow.
> 
> I'm aware that it's not officially supported (yet), it's part of a POC @nlwmn is working on to help AWS Customers to migrate CentOS 7 Systems to RHEL 7 PAYG. Manually and with Ansible playbooks Nicolas wrote, we did a lot successful conversions during the last month already.
> 

It's exciting to hear that you're trying to make the conversions work with PAYG instances on AWS! Would you mind sharing the playbooks?

Are you trying the conversion with just the free CentOS 7 images where one pays just for the EC2 infrastructure, or also images with a software cost being paid to a 3rd party vendor (such as [Center for Internet Security](https://aws.amazon.com/marketplace/seller-profile?id=dfa1e6a8-0b7b-4d35-a59c-ce272caee4fc))? What I found is that when converting the latter, the software cost is still being charged even after the conversion to RHEL because the software cost is associated with the image AMI and that doesn't change.

When converting a free CentOS 7 instance to RHEL 7 PAYG, how do you make sure that AWS sees the instance as RHEL 7 and bills the user appropriately?

If you'd rather take these topics off GitHub, let me know and I'll ping you on Matrix.

File: tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py (hunk 1 above)
Comment (danmyway) [reply tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:46]: It would timeout on the default 30 seconds

File: tests/ansible_collections/roles/add-custom-repos/rhel7-repos.yml [hunk 2]
```diff
@@ -4,7 +4,7 @@
       yum_repository:
         name: rhel-7-server-rpms
         description: RHEL 7.9 for $basearch
-        baseurl: https://rhsm-pulp.corp.redhat.com/content/dist/rhel/server/7/$releasever/$basearch/os/
+        baseurl: https://rhsm-pulp.corp.redhat.com/content/dist/rhel/server/7/7Server/$basearch/os/
```
[the remark this answers is not in the document]
Comment (danmyway) [reply tests/ansible_collections/roles/add-custom-repos/rhel7-repos.yml:7]: the `$releasever` gets resolved to `7` instead of `7Server`

File: tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py (hunk 1 above)
Comment (kokesak) [reply tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:46]: alright, cool

**Comment by bocekm** [comment]
Rebased. I'll check now if unit test updates are needed.

**Comment by wombelix** [comment]
@bocekm thanks for actively working on this PR, to your questions:

> It's exciting to hear that you're trying to make the conversions work with PAYG instances on AWS! Would you mind sharing the playbooks?

It's an early stage and not clear yet if it will become a stable and scalable solution, as soon we know for sure that it will get beyond the POC state we will definitely share more details.

> Are you trying the conversion with just the free CentOS 7 images where one pays just for the EC2 infrastructure, or also images with a software cost being paid to a 3rd party vendor

The focus is on free Marketplace AMIs

> What I found is that when converting the latter, the software cost is still being charged even after the conversion to RHEL because the software cost is associated with the image AMI and that doesn't change.

That's a correct observation, the relevant product code id is linked to the EBS and even if you move this EBS to another instance or change things on OS level it will still be charged in that scenario

> When converting a free CentOS 7 instance to RHEL 7 PAYG, how do you make sure that AWS sees the instance as RHEL 7 and bills the user appropriately?

A free Marketplace AMI doesn't have a product code or billing id attached. Our approach is to deploy a new RHEL 7 PAYG instance, which has the billing id associated with the actual ec2 instance and not the EBS volume. Then we swap the root volume, we attach the volume with the CentOS 7 to the RHEL 7 instance. In that case the billing code of the ec2 will still be used which makes the CentOS 7 system from a billing perspective to a RHEL 7 PAYG.

I already pinged you by mail and introduced you to Nicolas, the POC is his baby and he can share further insights with you. Would be great to move this initiative forward together, we things it's a very interesting use case for convert2rhel.

File: tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    shell("yum-config-manager --disable *")
+    # Enable just the rhel-7-server-rpms repo
+    shell("yum-config-manager --enable rhel-7-server-rpms --releasever 7Server")
+
+    with convert2rhel(
+        "-y --no-rpm-va --serverurl {} --username {} --password {} --pool {} --debug".format(
+            env.str("RHSM_SERVER_URL"),
+            env.str("RHSM_USERNAME"),
+            env.str("RHSM_PASSWORD"),
+            env.str("RHSM_POOL"),
```
Comment (r0x0d) [inline tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:43]: Out of curiosity, this will interfere with the changes done for the SCA PR?  In there, we removed all `RHSM_POOL` usage.

Comment (bocekm) [reply tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:43]: @r0x0d, it won't interfere.

Using the pool ID will still be possible, it just won't be required. Or, to put it more precisely - before the @pr-watson's PR https://github.com/oamg/convert2rhel/pull/814 the user needed to pass either `--pool` or  `--auto` when using the RHSM username and password. With the @pr-watson's PR they can just pass the username and password and that's it. But if they want they can still use the `--pool` option and it will work.

Comment (danmyway) [reply tests/integration/tier1/kernel-check-skip/test_latest_kernel_check_skip.py:43]: I would keep this as the one scenario using the `--pool` options.
Other cases will utilize the changes from #814

**Review by SpyTec — APPROVED** [review]

**Comment by SpyTec** [comment]
Thanks for helping the tool! Merged
